Documentation/en-us/DataDrivenExamples.md

```diff
@@ -0,0 +1,194 @@
+                .init(input: 2, output: 4),
+            ]
+            dataSet.forEach { (data) in
+                context("for input \(data.input)") {
+                    it("should return \(data.output)", line: data.line) {
+                        expect(line: data.line, double(data.input)).to(equal(data.output))
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+## More dimensions and complex test data
+
+Should there be multiple aspects or combinations of input to test or results that
+could be multi-dimensional, you can of course extend the tests to suit that.
+
+Here is an example where a class, `WeirdClass` should be created from a string,
+`init(String)`, and have two different derived properties, `weirdness` and `firstThree`.
+
+*Note that the tracing of line numbers is left out in this example for conciseness.*
+
+```swift
+import Quick
+import Nimble
+
+class WeirdClassTests: QuickSpec {
+    override func spec() {
+        describe("the WeirdClass") {
+            let dataSet: [(input: String, firstThree: String, weirdness: Int)] = [
+                (input: "foobar", firstThree: "foo", weirdness: 35),
+                (input: "Hello world", firstThree: "Hel", weirdness: 524),
+                (input: "78ghkajsdf", firstThree: "78g", weirdness: 1240),
+            ]
+            dataSet.forEach { (input, firstThree, weirdness) in
+                context("for input \(input)") {
+                    var weird: WeirdClass!
+                    beforeEach {
+                      weird = WeirdClass(input)
+                    }
+
+                    it("should have weirdness \(output)") {
+                      expect(weird.weirdness).to(equal(weirdness))
+                    }
+                    it("shoud have first three \(firstThree)") {
+                      expect(weird.firstThree).to(equal(firstThree)
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+## Further reducing the boilerplate
+
+Define a little generic helper function called `given` like this:
+
+```swift
+/**
+ Used to iterate through a set of given data. This can then be used to provide multiple examples of data.
+- Parameter input: The list of data to iterate through.
+- Parameter then: The closure to execute each item on.
+ */
+func given<Input>(_ input: Input..., then: (_: Input) -> Void) {
+    for i in input { then(i) }
+}
+```
+
+With that function you can write your specification like:
+
+```swift
+given(
+    (1, plus: 1, is: 2),
+    (1, plus: 2, is: 2),
+    (1, plus: 3, is: 4),
+) { (a, b, result) in
+    it("\(a) plus \(b) is \(result)") {
+        expect(a + b).to(equal(result))
+    }
+}
+```
+
+### Tracing the line on errors
+
+Should you want Xcode to trace the line when tests fail, you can add this, like so:
+
+```swift
+given(
+    (1, plus: 1, is: 2, line: #line),
+    (1, plus: 2, is: 2, line: #line),
+    (1, plus: 3, is: 4, line: #line),
+) { (a, b, result, line) in
```

Comment:
This should be `line: UInt` to compile correctly.

Reply:
Actually it is correctly inferred as `UInt`.

Reply:
Hmm, ok. That is what I originally expected, but hasn't always been my experience in practice.